File: backend/Core/Models/Dtos/Database/FlowAreaDto.cs
using Core.Enums;

namespace Core.Models.Dtos
{
    public class FlowAreaDto
    {
        /// <summary>0 for a new row. Negative for a new row another new row points at.</summary>
        public int Id { get; set; }

        public string Name { get; set; } = string.Empty;
        public FlowAreaTypeEnum Type { get; set; }


        // CUSTOM
        public int? ParentFlowAreaId { get; set; }
        public AreaSizingModeEnum SizingMode { get; set; }

        public int LocationX { get; set; }
        public int LocationY { get; set; }
        public int Width { get; set; }
        public int Height { get; set; }

        public float RatioX { get; set; }
        public float RatioY { get; set; }
        public float RatioWidth { get; set; }
        public float RatioHeight { get; set; }


        // APPLICATION, BROWSER_TAB
        public string ProcessName { get; set; } = string.Empty;
        public string TitlePattern { get; set; } = string.Empty;
        public TitleMatchModeEnum TitleMatchMode { get; set; }
        public bool UseClientArea { get; set; } = true;


        // BROWSER_TAB
        public string TabMatchValue { get; set; } = string.Empty;
        public TabMatchOnEnum TabMatchOn { get; set; }


        // MONITOR
        public string MonitorUniqueId { get; set; } = string.Empty;


        public int FlowId { get; set; }

        // Read only, projected by the query.
        public int FlowStepsCount { get; set; }
        public string ParentName { get; set; } = string.Empty;
    }
}

File: frontend/src/shared/models/database/flow-area-dto.ts
import { FlowAreaTypeEnum } from "@/shared/enums/backend/flow-area-type.enum";
import { AreaSizingModeEnum } from "@/shared/enums/backend/area/area-sizing-mode-enum";
import { TitleMatchModeEnum } from "@/shared/enums/backend/area/title-match-mode-enum";
import { TabMatchOnEnum } from "@/shared/enums/backend/area/tab-match-on-enum";
import type { ScalesWithEnum } from "@/shared/enums/backend/area/scales-with-enum";

export class FlowAreaDto {
  // 0 for a new row. New rows get a negative id so a sibling added in the same save can
  // reference them; the backend swaps them for real ids after the insert.
  id: number = 0;
  name: string = "";
  type: FlowAreaTypeEnum = FlowAreaTypeEnum.CUSTOM;

  // What makes its contents bigger or smaller on another screen. Null inherits the parent's, and
  // an area with no parent is DPI.
  scalesWith?: ScalesWithEnum | null;
  // The DPI its pixel numbers were written at.
  authoredDpi: number = 0;

  // CUSTOM
  parentFlowAreaId?: number | null;
  sizingMode: AreaSizingModeEnum = AreaSizingModeEnum.ABSOLUTE_PX;

  locationX: number = 0;
  locationY: number = 0;
  width: number = 0;
  height: number = 0;

  ratioX: number = 0;
  ratioY: number = 0;
  ratioWidth: number = 0;
  ratioHeight: number = 0;

  // APPLICATION, BROWSER_TAB
  processName: string = "";
  titlePattern: string = "";
  titleMatchMode: TitleMatchModeEnum = TitleMatchModeEnum.CONTAINS;
  useClientArea: boolean = true;

  // BROWSER_TAB
  tabMatchValue: string = "";
  tabMatchOn: TabMatchOnEnum = TabMatchOnEnum.TITLE;

  // MONITOR. Empty is the primary monitor.
  monitorDeviceName: string = "";

  flowId: number = 0;

  // Read only, set by the backend.
  flowStepsCount: number = 0;
  parentName: string = "";

  constructor(data: Partial<FlowAreaDto> = {}) {
    Object.assign(this, {
      ...data,
    });
  }
}
